[CBRD-25659] add options to leave heap profile data on JVM OutOfMemory error

hyunikn opened the pull request:
<http://jira.cubrid.org/browse/CBRD-25659>

문제 현상이 재현이 안되고 있습니다. 
현상이 발생했을 때 원인 분석을 할 수 있도록 heap dump 와 로그를 남기는 수정을 하였습니다. 
정상 실행에 있어 성능 저하는 없다고 볼 수 있습니다. 

- JVM OutOfMemory error 가 발생했을 때 heap profile data 를 남겨 원인 분석이 가능하도록 JVM 옵션 두 가지를 추가.
  - -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=jvmheapdump.hprof
- jvmheapdump.hprof 가 남았을 때 보관하도록 CI 설정 파일 수정
- PL 서버가 시작할 때 다음 내용을 로그 파일에 남김
  - 적용된 JVM 옵션
  - 실제 할당된 heap max size (Runtime.getRuntime().maxMemory() 로 얻어짐)

hyunikn committed:
log max heap size and JVM options on PL server starts

hyunikn committed:
add two JVM options to profile heap memory on OutOfMemory error

hyunikn committed:
temporary change to cause errors on test_plcsql test

hyunikn committed:
update circle ci config to copy jvmheapdump.hprof (if any) to /tmp/log directory

hyunikn committed:
limit JVM max heap size to 500m for temporary test

hyunikn committed:
temporary

hyunikn committed:
temporary

hyunikn committed:
Merge branch 'develop' into cbrd-25659-tackle-oom-error

hyunikn committed:
restore a temporary change which caused CI failures

hyunikn committed:
minor blank

hgryoo commented on src/sp/pl_sr_jvm.cpp:
```diff
@@ -447,6 +447,10 @@ pl_jvm_options ()
 
+  // CBRD-25659: dump heap memory on JVM OutOfMemory error
+  options.push_back ("-XX:+HeapDumpOnOutOfMemoryError");
+  options.push_back ("-XX:HeapDumpPath=jvmheapdump.hprof");
```
HeapDumpPath의 경로를 $CUBRID/log 또는 $CUBRID/tmp 디렉토리로 설정하면 좋을 것 같습니다.

hyunikn replied:
네. 반영하겠습니다.

hyunikn committed:
refinements
  . guarantee that closeSocket() is run by putting it in finally block
  . guarantee the the socket is closed when an exception occurs before ExecThread is run
  . put jvmheapdump.hprof in log directory

hyunikn commented:
추가 수정이 있었습니다. 
- ExecThread 의 에러 때에도 socket이 close 되도록 보강 - 매우 작은 확률로 문제의 소지가 있는 듯 하여 수정했습니다. 
    - ExecThread.run() 안에서 closeSocket() 을 finally 안에 집어 넣어 반드시 실행되게 하기
    - ListenerThread 안에서 ExecuteThread 가 run 되기 전에 exception 으로 실패할 때 socket close 보장
-  jvmheapdump.hprof 가 log 디렉토리 아래에 생기도록 조정
    - circleci 설정에서 jvmheapdump.hprof 복사해 주던 것을 하지 않음 - log 들과 함께 복사될 것이므로.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/jsp/ExecuteThread.java:
```diff
@@ -130,145 +130,152 @@ public void closeSocket() {
-                            ctx = null;
-                            break;
-                        }
+        try {
```
여기서부터 275 라인까지는 indent 외에 다른 변경은 없습니다. 
277 라인의 closeSocket() 을 finally 블럭에 넣기 위한 변경입니다.

kangmin5505 replied:
추후에 비슷한 수정이 필요할 경우, 다음과 같은 방법도 추천드립니다.
[java 1.7부터 지원되는 try-with-resources](https://docs.oracle.com/javase/7/docs/technotes/guides/language/try-with-resources.html) 구문을 활용하시면, 자원 해제를 잊지 않고 편리하게 할 수 있습니다.

```java
// 자원을 해제해야 하는 클래스에 java.lang.AutoCloseable 인터페이스를 상속받아 close() 메서드를 override
try ( //해제할 자원1,
//해제할 자원2
// ...
) {
} catch {
// ...
}
```

hyunikn committed:
minor comment
